tests/testing-resources/include/aws/testing/AwsTestHelpers.h

```diff
@@ -68,4 +69,141 @@ OutcomeT CallOperationWithUnconditionalRetry(const ClientT* client,
+                {
+                    _suite.push_back(item);
+                    status = true;
+                }
+            }
+            AWS_LOGSTREAM_TRACE(ALLOCATION_TAG, "added function with id=" << item.entryId<<" num tries="<<item.retryCount<<" stop on fail="<<item.stopOnFail<<std::endl );
+            //std::cout<<"added function with id=" << item.entryId<<std::endl;
+            return status;
+        }
+
+        /*
+        this function executes the functional blocks in order according to user specification
+
+        One can execute from the beginning with appropriate application setup in the context or just resume from after last block executed
+
+        if retry is set for a block, then block will be retried for specified number of times on failure, whereas
+        in case of successful execution, the block will not be tried again
+
+        if stop on fail is set and a block fails even after retries, the execution of test suite ends there
+
+        */
+        bool execute(bool startFromBegin = false)
+        {
+            bool status = !_suite.empty();
+            auto it = _suite.begin();
+            
+            if(!startFromBegin)
+            {
+                std::advance(it, _cursor) ;
+            }
+            else 
+            {
+                _cursor = 0;
+            }
+
+            for(; it != _suite.end();++it)
+            {
+                if(!it->candidate)
+                {
+                    //std::cout<<"invalid function with id=" << it->entryId<<std::endl;
+                    return false;
+                }
+
+                {
+                    //backup count
+                    auto originalRetryCount = it->retryCount;
+                    bool localResult = false;
+
+                    //Try and retry a block
+                    do{}
+                    while( it->retryCount && !(localResult = it->candidate(_contextSp, it->entryId, --it->retryCount) ));
+                    
+                    //set global status from block evaluation
+                    status = status && localResult;
+
+                    //if stop on fail after retries, break and retry from start if available
+                    if(it->stopOnFail && !localResult)
+                    {
+                        break;
+                    } 
+                    //restore state
+                    it->retryCount = originalRetryCount;
+                }  
+                ++_cursor;            
+            }
+
+            AWS_LOGSTREAM_TRACE(ALLOCATION_TAG, "execute overall status="<<status );
+            //std::cout<<"execute overall status="<<status<<std::endl;
+
+            return status;
+        }
+
+        /*
+        this function clears all the functional blocks added and resets the cursor
+        */
+        void clear()
+        {
+            _suite.clear();
+            _cursor = 0;
+        }
+    private:
+        using FunctionSuite = std::list<FunctionBlock>;
```

Comment:
We can use any container, I used generic iterator interface for this reason. Can make this a template parameter with type traits checks. This is just enhancement